facebookads/api.py

```diff
@@ -500,25 +500,26 @@ def execute(self):
         retry_indices = []
 
         for index, response in enumerate(responses):
-            if response:
-                body = response.get('body')
-                code = response.get('code')
-                headers = response.get('headers')
-
-                inner_fb_response = FacebookResponse(
-                    body=body,
-                    headers=headers,
-                    http_status=code,
-                    call=self._batch[index],
-                )
-
-                if inner_fb_response.is_success():
-                    if self._success_callbacks[index]:
-                        self._success_callbacks[index](inner_fb_response)
-                elif self._failure_callbacks[index]:
-                    self._failure_callbacks[index](inner_fb_response)
-            else:
+            if not response:
                 retry_indices.append(index)
+                continue
+
+            inner_fb_response = FacebookResponse(
+                body=response.get('body'),
+                headers=response.get('headers'),
+                http_status=response.get('code'),
+                call=self._batch[index],
+            )
+            if inner_fb_response.is_success() and self._success_callbacks[index]:
+                self._success_callbacks[index](inner_fb_response)
+            else:
+                # retry transient errors
+                if inner_fb_response.is_transient():
+                    retry_indices.append(index)
+
+                # execute failure callbacks on error, even if transient
+                if self._failure_callbacks[index]:
```

Comment:
This is indeed another way of implementing it, and it does maintain the original signatures! However, it still doesn't give the callers the flexibility to deal with the added scenario differently.

I see your point of adding a transient_error_callback; we're basically assuming people are using the explicit parameter passing in FacebookAdsApiBatch.add, or it'll be a breaking change for them, or unless we add the transient_error_callback to the end of the parameters list. (the FacebookAdsApiBatch.add_request will be fine; we just add the transient_error to the end)

```python
def add(
        self,
        method,
        relative_path,
        params=None,
        headers=None,
        files=None,
        success=None,
        failure=None,
       transient_error=None, #<========= adding here
        request=None,
       transient_error=None, #<========= or adding here
    ):
```
I still prefer adding a transient_error_callback; it's more flexible and thus adding more value to the SDK in the long run. 

We could add a small check in the "add" method; like, testing if "transient_error" parameter type is request object; if it is, we could warn/err out/exception out since it's clearly the caller is using the non-explicit for optional parameters and the new code will break their logic.